stdlib/public/core/Span/Span.swift

```diff
@@ -0,0 +1,856 @@
+  /// an element of type `Element`, and must contain a number of bytes
+  /// that is an exact multiple of `Element`'s stride.
+  ///
+  /// - Parameters:
+  ///   - buffer: a buffer to initialized elements.
+  @_disallowFeatureSuppression(NonescapableTypes)
+  @_alwaysEmitIntoClient
+  //FIXME: should be @lifetime(borrow <argname>) rdar://138672380
+  @lifetime(immortal)
+  public init(
+    _unsafeBytes buffer: borrowing Slice<UnsafeMutableRawBufferPointer>
+  ) {
+    self.init(_unsafeBytes: UnsafeRawBufferPointer(rebasing: buffer))
+  }
+
+  /// Create a `Span` over the bytes represented by a `RawSpan`
+  ///
+  /// - Parameters:
+  ///   - rawSpan: An existing `RawSpan`, which will define both this
+  ///             `Span`'s lifetime and the memory it represents.
+  @_disallowFeatureSuppression(NonescapableTypes)
+  @_alwaysEmitIntoClient
+  @lifetime(rawSpan)
+  public init(_bytes rawSpan: consuming RawSpan) {
+    self.init(
+      _unsafeBytes: .init(start: rawSpan._pointer, count: rawSpan.byteCount)
+    )
+  }
+}
+
+@_disallowFeatureSuppression(NonescapableTypes)
+@available(SwiftStdlib 6.1, *)
+extension Span where Element: Equatable {
+
+  /// Returns a Boolean value indicating whether this and another span
+  /// contain equal elements in the same order.
+  ///
+  /// - Parameters:
+  ///   - other: A span to compare to this one.
+  /// - Returns: `true` if this sequence and `other` contain equivalent items,
+  ///   using `areEquivalent` as the equivalence test; otherwise, `false.`
+  ///
+  /// - Complexity: O(*m*), where *m* is the lesser of the length of the
+  ///   sequence and the length of `other`.
+  @_disallowFeatureSuppression(NonescapableTypes)
+  @_alwaysEmitIntoClient
+  public func _elementsEqual(_ other: Self) -> Bool {
+    guard count == other.count else { return false }
+    if count == 0 { return true }
+
+    // This could be short-cut with a layout constraint
+    // where stride equals size, as long as there is
+    // at most 1 unused bit pattern, e.g.:
+    // if Element is BitwiseEquatable {
+    // return _swift_stdlib_memcmp(lhs.baseAddress, rhs.baseAddress, count) == 0
+    // }
+    if _pointer != other._pointer {
+      for o in 0..<count {
+        if self[unchecked: o] != other[unchecked: o] { return false }
+      }
+    }
+    return true
+  }
+
+  /// Returns a Boolean value indicating whether this span and a Collection
+  /// contain equal elements in the same order.
+  ///
+  /// - Parameters:
+  ///   - other: A Collection to compare to this span.
+  /// - Returns: `true` if this sequence and `other` contain equivalent items,
+  ///   using `areEquivalent` as the equivalence test; otherwise, `false.`
+  ///
+  /// - Complexity: O(*m*), where *m* is the lesser of the length of the
+  ///   sequence and the length of `other`.
+  @_disallowFeatureSuppression(NonescapableTypes)
+  @_alwaysEmitIntoClient
+  public func _elementsEqual(_ other: some Collection<Element>) -> Bool {
```

Comment:
Why do we need to provide these `Sequence`/`Collection`-based methods?

Reply:
They're useful for writing the tests, at least. I will move them to the test file in the tests PR.